File: BinarySearchProject/binarySearchTree.cpp
/*
 * binarySearchTree.cpp
 *
 *  Created on: Jan 31, 2016
 *      Author: allen
 */


/**
*	Allen Bui
*	January 31, 2016
*	Binary Search Tree in C++
**/

#include "binarySearchTree.hpp"
using namespace std;

binarySearchTree::binarySearchTree() {
 	root = NULL;
}

void binarySearchTree::destroyTree() {
	destroyTreeUtil(root);
}

void binarySearchTree::destroyTreeUtil(node *leaf) {
	if (leaf == NULL) return;

	destroyTreeUtil(leaf->left);
	destroyTreeUtil(leaf->right);

	delete(leaf);
}

void binarySearchTree::insertNode(int key) {
	if (root != NULL) {
		root = new node;
		root->data = key;
		root->left = NULL;
		root->right = NULL;
	}
	else {
		insertNodeUtil(key, root);
	}
}

void binarySearchTree::insertNodeUtil(int key, node *leaf) {
	if (leaf == NULL) {
		leaf = new node;
		leaf->data = key;
		leaf->left = NULL;
		leaf->right = NULL;
		return;
	}
	else if (leaf->data == key) {
		return;
	}
	else if (leaf->data > key) {
		insertNodeUtil(key, leaf->left);
	} else {/*(leaf->data > key)*/
		insertNodeUtil(key, leaf->right);
	}
}

//depth first search
node *binarySearchTree::search(int key) {
	if (root == NULL) {
		return NULL;
	}
	else {
		return searchUtil(key, root);
	}
}

node *binarySearchTree::searchUtil(int key, node*leaf) {
	if (leaf == NULL) return NULL;
	if (leaf->data == key) return leaf;
	else if (leaf->data < key) searchUtil(key, leaf->right);
	else searchUtil(key, leaf->left);
}

node *binarySearchTree::deleteNode(int data) {
	return deleteNodeUtil(root, data);
}

node *binarySearchTree::deleteNodeUtil(node *leaf, int data) {
	if (leaf == NULL) return leaf;
	if (leaf->data < data) {
		deleteNodeUtil(leaf->right, data);
	}
	else if (leaf->data > data) {
		deleteNodeUtil(leaf->left, data);
	}
	else {	//found the node to delete
		int numChildren = evalNumChildren(leaf);

		if (numChildren == 0) {
			delete(leaf);
			leaf = NULL;
		}
		else if (numChildren == 1) {
			node *temp = leaf;
			if (leaf->left == NULL) {	//then it has a rightChild
				leaf = leaf->right;
				delete(temp);
			}
			else {//it has a leftChild
				leaf = leaf->left;
				delete(temp);
			}
		}
		else { //numChildren == 2
			node *successor = getSuccessor(leaf->right);
			leaf->data = successor->data;
			leaf->right = deleteNodeUtil(leaf->right, successor->data);
		}
	}
	return leaf;
}

node *binarySearchTree::getSuccessor(node *child) {
	if (child->left == NULL) {
		return child;
	}
	getSuccessor(child->left);
}

int binarySearchTree::evalNumChildren(node *leaf) {
	int numChildren = 0;

	if (leaf->right != NULL) {
		if (leaf->left != NULL) {
			numChildren = 2;
		}
		numChildren = 1;
	}

	if (leaf->left != NULL) {
		if (leaf->right != NULL) {
			numChildren = 2;
		}
		numChildren = 1;
	}

	return numChildren;
}


void binarySearchTree::inOrderTraversal() {
	inOrderTraversalUtil(root);
}

void binarySearchTree::inOrderTraversalUtil(node * leaf) {
	if(leaf == NULL) return;
	inOrderTraversalUtil(leaf->left);
	inOrderTraversalUtil(leaf->right);
	cout << leaf->data << endl;
}

/***********************************************************************************
*		utility functions to check whether students created correct tree structures
**/
bool binarySearchTree::isBinarySearchTree(node *leaf) {
	checkForBinarySearchTree(leaf, INT_MAX, INT_MIN);
}


//O(n)
bool binarySearchTree::checkForBinarySearchTree(node *leaf, int minValue, int maxValue) {
	if (leaf == NULL) return true;

	if (leaf->data > minValue && leaf->data < maxValue && checkForBinarySearchTree(leaf->left, minValue, leaf->data) && checkForBinarySearchTree(leaf->right, leaf->data, maxValue)) return true;
	else return false;
}

bool binarySearchTree::isSubTreeLesser(node *leaf, int data) {
	if (leaf == NULL) return true;
	if (leaf->data <= data && isSubTreeLesser(leaf->left, data) && isSubTreeLesser(root->right, data)) return true;
	else return false;
}


bool binarySearchTree::isSubTreeGreater(node *leaf, int data) {
	if (leaf == NULL) return true;
	if (leaf->data  < data && isSubTreeGreater(leaf->left, data) && isSubTreeGreater(root->right, data)) return true;
	else return false;
}

/*  NAIVE APPROACH

//O(n^2)
bool checkForBinarySearchTree(node *leaf) {

	if (leaf == NULL) return true;

	if (isSubTreeLesser(leaf->left, leaf->data) && isSubTreeGreater(leaf->right, leaf->data) && checkForBinarySearchTree(root->left) && checkForBinarySearchTree(root->right)) return true;
	else {return false};
}

*/

File: CubicPermutations/src/BinarySearchTree.java
import java.util.ArrayList;


public class BinarySearchTree {

	public static Node root;
	
	public BinarySearchTree() {
		root = null;
	}
	
	public Node getRoot() {
		return root;
	}
	
	public boolean findNodeByValue(int val) {
		Node current = Node.deepCopy(root);
		
		while (current != null) {
			if (current.value == val) {
				return true;
			} else if (current.value >= val) {
				current = current.left;
			} else {
				current = current.right;
			}
		}
		return false;
	}
	
	public boolean findNodeByCount(int val) {
		
		Node current = Node.deepCopy(root);
		
		while (current != null) {
			if (current.count == 3) {
				return true;
			} else if (current.value >= val) {
				current = current.left;
			} else {
				current = current.right;
			}
		}
		return false;
	}
	
	public void deleteNode(int val) {
		deleteNodeUtil(root, val);
	}
	
	private void deleteNodeUtil(Node node, int val) {
		
		if (root == null) {
			return;
		}
		
		if (val < node.value) {
			
			deleteNodeUtil(node.getLeft(), val);
			
		} else if ( val > node.value){
			
			deleteNodeUtil(node.getRight(), val);
			
		} else {		//found the node

			Node successor = node.getRight();
			
			if (node.getLeft() != null && node.getRight() != null) { // two children in the tree
				if (successor != null) {
					
					successor = getSuccessor(successor);
					node.setValue(successor.getValue());
					
				}
				
			} else if (node.getLeft() == null && node.getRight() != null) { // node has a right child
				
				if (node.getParent().getRight() == node) {		// node is a rightChild of parent
					
					node.getParent().setRight(node.getRight());
					node.getRight().setParent(node.getParent());
					
				} else {										//node is a leftChild of a parent
					
					node.getParent().setLeft(node.getRight());
					node.getRight().setParent(node.getParent());
					
				}
				
			} else if (node.getLeft() != null && node.getRight() == null) { // node has a left child
				
				if (node.getParent().getRight() == node) { // node is a rightChild
					
					node.getParent().setRight(node.getLeft());
					node.getLeft().setParent(node.getParent());
					
				} else {
					
					node.getParent().setLeft(node.getLeft());
					node.getLeft().setParent(node.getParent());
					
				}
			} else { // node has no children
				
				if( node.getParent().getRight() == node) { //the node is the rightChild of the parent
					
					node.getParent().setRight(null);
					
				} else { // the node is a leftChild of the parent
					node.getParent().setRight(null);
				}
			}
		}					
	}
	
	private Node getSuccessor(Node successor) {
		if (successor != null) {
			successor = successor.getLeft();
		}
		return successor;
	}
	
	
	
	public void insertNode(long val) {		
		insertNodeUtil(root, val);
	}
	
	private void insertNodeUtil(Node tree, long val) {
		
		if (tree == null) {
			
			tree = new Node(tree, val);
			
		}
		
/*		if (tree.getValue() == val) {
			
			tree.incCount();
			return;
			
		}*/
		
		if (tree.getValue() > val) {
			
			if (tree.getLeft() == null) {
				
				tree.setLeft(new Node(tree, val));
				
			} else {
				
				insertNodeUtil(tree.getLeft(), val);
				
			}
			
		}
		
		if (tree.getValue() < val) {
			
			if (tree.getRight() == null) {
				tree.setRight(new Node(tree, val));
			} else {
				insertNodeUtil(tree.getRight(), val);	
			}
		}
	}
	
	public String toString() {
		
		toStringUtil(root);
		return "done";
		
	}
	
	private static void toStringUtil( Node node ) {
		if (node != null) {
			toStringUtil(node.getLeft());
			System.out.println(node.getValue());
			toStringUtil(node.getRight());
		}
	}
}
